notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala

```diff
@@ -15,237 +15,162 @@ import java.io.File
+
+  /** @return the content of the file present at `location` as an array of bytes if there is one */
+  def readBytes(location: Location): LocationIO[Array[Byte]] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => fromRIO(p.readBytes.flatMap(_.cata(RIO.ok, RIO.fail(s"Failed to read file - LocalPath(${p.path}) does not exist"))))
+      case HdfsLocationContext(conf, p) => fromHdfs(conf, p.readBytes.flatMap(_.cata(Hdfs.ok, Hdfs.fail(s"Failed to read file - HdfsPath(${p.path}) does not exist"))))
+      case S3LocationContext(client, p) => fromS3(client, p.getBytes)
     }
 
-  /** pipe bytes from one location to another */
-  def pipe(from: Location, to: Location): RIO[Unit] =
-    readUnsafe(from)(in => writeUnsafe(to)(out => Streams.pipe(in, out)))
+  def readUnsafe(location: Location)(f: java.io.InputStream => LocationIO[Unit]): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => LocationIO(ioc => p.determineFile.flatMap(l => RIO.using(l.toInputStream)(i => f(i).run(ioc))))
+      case HdfsLocationContext(conf, p) => LocationIO(ioc => p.readWith(i => Hdfs.fromRIO(f(i).run(ioc))).run(conf))
+      case S3LocationContext(client, p) => LocationIO(ioc => p.withStream(i => f(i).run(ioc)).run(client).map(_._2)) // TODO drop AwsLog?
+    }
 
-  /**
-   * synchronize files from one location to another.
-   *
-   * if the first location is a directory, copy all the files in that directory to the other location (must be a directory as well)
-   *
-   * files existing at destination are not being copied
-   *
-   * Note: in the case of a large copy from S3 to Hdfs or from Hdfs to S3, a distcopy should be done instead
-   * see SynchronizedInputsOutputs
-   *
-   * @return the list of copied locations
-   */
-  def syncFiles(from: Location, to: Location): RIO[List[Location]] =
-    isDirectory(from) >>= { isDirectory =>
-      if (isDirectory)
-        list(from).flatMap(_.traverseU {
-          case f @ LocalLocation(_) => syncFile(f, to </> f.filePath).as(f: Location)
-          case f @ HdfsLocation(_)  => syncFile(f, to </> f.filePath).as(f: Location)
-          case s @ S3Location(_, k) => syncFile(s, to </> FilePath.unsafe(k)).as(s: Location)
-        })
-      else
-        syncFile(from, to).as(List(from))
+  /** write to a given location, using an OutputStream */
+  def writeUnsafe(location: Location)(f: java.io.OutputStream => LocationIO[Unit]): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p) => LocationIO(ioc => for {
+        _ <- p.doesNotExist(s"Can not overwrite location ${location}", p.dirname.mkdirs)
+        r <- RIO.using(p.path.toOutputStream)(o => f(o).run(ioc))
+      } yield r)
+      case HdfsLocationContext(conf, p) => LocationIO(ioc => p.writeWith(o => Hdfs.fromRIO(f(o).run(ioc))).run(conf))
+      case S3LocationContext(client, p) => LocationIO(ioc => for {
```

Comment:
Can we just use [putStream](https://github.com/ambiata/saws/blob/master/saws-s3/src/main/scala/com/ambiata/saws/s3/S3Pattern.scala#L149) and map here?